src/hotspot/share/opto/phaseX.cpp

```diff
@@ -836,21 +836,25 @@ Node *PhaseGVN::transform( Node *n ) {
 //------------------------------transform--------------------------------------
 // Return a node which computes the same function as this node, but
 // in a faster or cheaper fashion.
-Node *PhaseGVN::transform_no_reclaim( Node *n ) {
+Node *PhaseGVN::transform_no_reclaim(Node *n) {
   NOT_PRODUCT( set_transforms(); )
 
   // Apply the Ideal call in a loop until it no longer applies
-  Node *k = n;
-  NOT_PRODUCT( uint loop_count = 0; )
-  while( 1 ) {
-    Node *i = apply_ideal(k, /*can_reshape=*/false);
-    if( !i ) break;
-    assert( i->_idx >= k->_idx, "Idealize should return new nodes, use Identity to return old nodes" );
+  Node* k = n;
+  Node* i = apply_ideal(k, /*can_reshape=*/false);
+  NOT_PRODUCT(uint loop_count = 1;)
+  while (i != NULL) {
+    assert(i->_idx >= k->_idx, "Idealize should return new nodes, use Identity to return old nodes" );
     k = i;
-    assert(loop_count++ < K, "infinite loop in PhaseGVN::transform");
+#ifdef ASSERT
+    if (loop_count >= K * C->live_nodes()) {
```

Comment:
This should be `K + C->live_nodes()` instead of `*`.
The code in `optimize()` uses `*` because it goes over all nodes. But here you iterate only one node.

Reply:
It's not very obvious to me. I think the limits should actually be the same everywhere because in the edge case, when all the nodes are connected and you start with the right one, all the same optimizations can happen within the loop in `transform_old`. Imagine a graph that is a huge foldable one-side-leaning expression tree. Its processing can either fully happen in the loop in `transform_old` if you start from the top, or it can happen in `optimize` when you start from the bottom and do the worklist-based processing one by one. Depends whether you can go forward in the graph or need to backtrack, but the amount of computation is kind of the same. Or am I wrong?

Reply:
Yes, for such particular graph number of computation will be about the same. But not in general case.
`N * 1000` is overkill in `transform_old()` - it will not help with catching bad cases if you have to wait so long.
And the limit in `optimize()` should not be the same - in worse cases it does `* N` computations done in `transform_old()`.